### PR TITLE
Separate gRPC server for internal/external usage

### LINKED ISSUE #1293: Separate gRPC server for internal/external calls
Today, Dapr uses the same gRPC server for both external (client to Dapr) and internal (Dapr to Dapr) calls.

Over time, different low level network policies might need to be applied differently.
One example is mTLS, where gRPC's behavior for `VerifyClientCertIfGiven` results in having clients needing to configure an empty TLS object when calling Dapr.

Other examples may include different middleware for internal and external calls.

### PR DESCRIPTION
**Pull request by yaron2**
Closes #1293 

This PR does the following:

1. Create separate gRPC servers for external and internal APIs
2. Sidecar TLS policy for internal server is set to require certificate, thereby strengthening security
3. Updates the sidecar injector and operators to mind the new port
4. For self-hosted mode, `daprd` will pick an available port if not given
5. Updated tests
6. Removed components handler from the operator, which was never functional and will be re-done as part of https://github.com/dapr/dapr/issues/1278
7. Does not require CLI changes

Validated on K8s and self-hosted modes.

### CODE REVIEW

File: pkg/grpc/port.go [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package grpc
+
+import "net"
+
+func GetFreePort() (int, error) {
```
Comment (youngbupark) [inline pkg/grpc/port.go]: Nit: It's a public method so it needs comments. linter will not be happy.

File: pkg/grpc/server.go [hunk 2]
```diff
@@ -130,7 +147,7 @@ func (s *server) getGRPCServer() (*grpc_go.Server, error) {
 
 		tlsConfig := tls.Config{
 			ClientCAs:  s.signedCert.TrustChain,
-			ClientAuth: tls.VerifyClientCertIfGiven,
+			ClientAuth: tls.RequireAndVerifyClientCert,
```
[the remark this answers is not in the document]
Comment (yaron2) [reply pkg/grpc/server.go]: If a user disables mTLS, the `authenticator` is nil and this line of code doesn't get executed.

File: pkg/grpc/port.go (hunk 1 above)
Comment (yaron2) [reply pkg/grpc/port.go]: The linter is happy, but you're right.

File: pkg/runtime/cli.go [hunk 3]
```diff
@@ -85,7 +87,20 @@ func FromFlags() (*DaprRuntime, error) {
...
 
-	applicationPort := 0
+	var daprInternalGRPC int
+	if *daprInternalGRPCPort != "" {
+		daprInternalGRPC, err = strconv.Atoi(*daprInternalGRPCPort)
+		if err != nil {
+			return nil, fmt.Errorf("error parsing dapr-internal-grpc-port: %s", err)
+		}
+	} else {
+		daprInternalGRPC, err = grpc.GetFreePort()
```
[the remark this answers is not in the document]
Comment (yaron2) [reply pkg/runtime/cli.go]: This is the same method used in the CLI

File: pkg/grpc/port.go (hunk 1 above)
Comment (yaron2) [reply pkg/grpc/port.go]: Updated now.

File: pkg/runtime/runtime.go [hunk 4]
```diff
@@ -500,14 +510,24 @@ func (a *DaprRuntime) startHTTPServer(port, profilePort int, allowedOrigins stri
 	server.StartNonBlocking()
 }
 
-func (a *DaprRuntime) startGRPCServer(port int) error {
-	api := grpc.NewAPI(a.runtimeConfig.ID, a.appChannel, a.stateStores, a.secretStores, a.getPublishAdapter(), a.directMessaging, a.actor, a.sendToOutputBinding, a)
+func (a *DaprRuntime) startgRPCInternalServer(api grpc.API, port int) error {
```
[the remark this answers is not in the document]
Comment (yaron2) [reply pkg/runtime/runtime.go]: Updated.

File: pkg/runtime/cli.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (yaron2) [reply pkg/runtime/cli.go]: Also, we do not randomize ports in other parts of `daprd`.

[the remark this answers is not in the document]
Comment (youngbupark) [reply pkg/runtime/cli.go]: > Also, we do not randomize ports in other parts of daprd.

There is no code to change port in daprd. CLI randomized the ports. so question is that why not randomizing the internal port in cli ?

[the remark this answers is not in the document]
Comment (yaron2) [reply pkg/runtime/cli.go]: Because this code run on Kubernetes mode as well.

[the remark this answers is not in the document]
Comment (yaron2) [reply pkg/runtime/cli.go]: Also, this is the only port that is not used externally by users. It's for Dapr to Dapr communications. the ports randomized by the CLI are all ports for user facing endpoints..

[the remark this answers is not in the document]
Comment (youngbupark) [reply pkg/runtime/cli.go]: > Also, this is the only port that is not used externally by users. It's for Dapr to Dapr communications. the ports randomized by the CLI are all ports for user facing endpoints..

make sense :)

File: pkg/grpc/port.go [hunk 5]
```diff
@@ -0,0 +1,18 @@
+package grpc
```
Comment (youngbupark) [inline pkg/grpc/port.go]: license is missing :)

File: pkg/grpc/port.go (hunk 1 above)
Comment (shalabhms) [inline pkg/grpc/port.go]: Could we use freeport.GetFreePort() which is getting already used in Dapr and get rid of this file?

File: pkg/grpc/server.go [hunk 6]
```diff
@@ -45,16 +47,28 @@ type server struct {
...
+	return &server{
+		api:         api,
+		config:      config,
+		tracingSpec: tracingSpec,
+		kind:        apiServer,
+	}
+}
+
+// NewInternalServer returns a new gRPC server for Dapr to Dapr communications
+func NewInternalServer(api API, config ServerConfig, tracingSpec config.TracingSpec, authenticator auth.Authenticator) Server {
```
Comment (shalabhms) [inline pkg/grpc/server.go]: Removed the nit comment - as I read more, NewInernalServer is fine.

File: pkg/grpc/port.go (hunk 5 above)
Comment (yaron2) [reply pkg/grpc/port.go]: Updated

File: pkg/grpc/port.go (hunk 1 above)
Comment (yaron2) [reply pkg/grpc/port.go]: It's used in tests, not in `daprd`. I want to limit the amount of external dependencies as much as possible for `daprd`.

**Pull request merged**